worker/firewaller/firewaller.go

```diff
@@ -1100,61 +1126,238 @@ func diffRanges(currentRules, wantedRules []network.IngressRule) (toOpen, toClos
 	localApplicationTag names.ApplicationTag
+	remoteRelationId    *params.RemoteEntityId
+	remoteModelUUID     string
 	networks            set.Strings
 }
 
-// relationLifeChanged looks up the required ingress CIDRs to allow
-// the specified relation to accept traffic from another model.
-// TODO(wallyworld) - listen for network changes as the networks may not be known
-// when the relation change is first noticed
-func (fw *Firewaller) relationLifeChanged(tag names.RelationTag) error {
-	rel, err := fw.firewallerApi.Relation(tag)
-	found := !params.IsCodeNotFound(err)
-	if found && err != nil {
-		return err
+// startRelation creates a new data value for tracking details of the
+// relation and starts watching the related models for subnets added or removed.
+func (fw *Firewaller) startRelation(rel *params.RemoteRelation) error {
+	tag := names.NewRelationTag(rel.Key)
+	data := &remoteRelationData{
+		fw:                  fw,
+		tag:                 tag,
+		remoteModelUUID:     rel.SourceModelUUID,
+		localApplicationTag: names.NewApplicationTag(rel.ApplicationName),
+		relationReady:       make(chan params.RemoteEntityId),
 	}
-	dead := !found || rel.Life() == params.Dead
+	fw.relationIngress[tag] = data
 
-	data, known := fw.applicationIngress[tag]
-	logger.Debugf("relation %v has existing data: %+v", tag, data)
+	err := catacomb.Invoke(catacomb.Plan{
+		Site: &data.catacomb,
+		Work: func() error {
+			return data.watchLoop()
+		},
+	})
+	if err != nil {
+		delete(fw.relationIngress, tag)
+		return errors.Trace(err)
+	}
 
-	change := false
-	if !known && dead {
-		return nil
+	// register the relationData with the firewaller's catacomb.
+	if err := fw.catacomb.Add(data); err != nil {
+		delete(fw.relationIngress, tag)
+		return errors.Trace(err)
 	}
-	if known && dead {
-		logger.Debugf("relation was known but has died")
-		delete(fw.applicationIngress, tag)
-		change = true
+
+	return fw.startRelationPoller(rel, data.relationReady)
+}
+
+// watchLoop watches the relation for networks added or removed.
+func (rd *remoteRelationData) watchLoop() error {
+	// TODO(wallyworld) - don't create one connection per relation
+	facade, err := rd.fw.newRemoteFirewallerAPIFunc(rd.remoteModelUUID)
+	if err != nil {
+		return errors.Trace(err)
 	}
-	if !known && !dead {
-		logger.Debugf("relation is new")
-		networks, err := fw.remoteRelationsApi.IngressSubnetsForRelation(tag.Id())
-		if err != nil {
+	defer facade.Close()
+
+	subnetsWatcher, err := facade.WatchSubnets()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for {
+		select {
+		case <-rd.catacomb.Dying():
+			// We stop the watcher here as it is tied to the remote facade
+			// which is closed as soon as we return.
+			worker.Stop(subnetsWatcher)
+			return rd.catacomb.ErrDying()
+		case <-subnetsWatcher.Changes():
+			if rd.remoteRelationId == nil {
+				// relation not ready yet.
+				continue
+			}
+			logger.Debugf("subnets changed in model %v", rd.remoteModelUUID)
```

Comment:
Wouldn't you need to do something here? Oh, I think I understand - we go around the loop again and that calls updateNetworks. But we don't bother if there's no `remoteRelationId` yet.